// tst-reflect/dist/descriptions/constructor-import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructorImportActivator = exports.ConstructorImport = void 0;
/**
 * Method details
 */
class ConstructorImport {
    /**
     * @internal
     */
    constructor(description) {
        if (new.target != ConstructorImportActivator) {
            throw new Error("You cannot create instance of Method manually!");
        }
        if (!description) {
            return;
        }
        this._name = description.n;
        this._exportName = description.en;
        this._sourcePath = description.srcPath;
        this._outPath = description.outPath;
    }
    /**
     * This is the name of the actual declaration
     * In the example above, this would be "SomeClass"
     */
    get name() {
        return this._name;
    }
    /**
     * The exported name of this constructor from its source file.
     * For example;
     * "export class SomeClass {}" would be "SomeClass"
     * "export default class SomeClass {}" would be "default"
     */
    get exportName() {
        return this._exportName;
    }
    /**
     * The absolute path of the source file for this constructor
     */
    get sourcePath() {
        return this._sourcePath;
    }
    /**
     * The absolute path for the javascript file of this constructor
     */
    get outPath() {
        return this._outPath;
    }
}
exports.ConstructorImport = ConstructorImport;
/**
 * @internal
 */
class ConstructorImportActivator extends ConstructorImport {
}
exports.ConstructorImportActivator = ConstructorImportActivator;
