// tst-reflect/dist/descriptions/propertyInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyInfoActivator = exports.PropertyInfo = void 0;
const enums_1 = require("../enums");
const mapper_1 = require("../mapper");
const Type_1 = require("../Type");
/**
 * Property description
 */
class PropertyInfo {
    /**
     * @param description
     * @internal
     */
    constructor(description) {
        var _a, _b, _c, _d;
        this.name = description.n;
        this._type = new Type_1.LazyType(description.t);
        this._decorators = ((_a = description.d) === null || _a === void 0 ? void 0 : _a.map(mapper_1.Mapper.mapDecorators)) || [];
        this.optional = description.o;
        this.accessModifier = (_b = description.am) !== null && _b !== void 0 ? _b : enums_1.AccessModifier.Public;
        this.accessor = (_c = description.acs) !== null && _c !== void 0 ? _c : enums_1.Accessor.None;
        this.readonly = (_d = description.ro) !== null && _d !== void 0 ? _d : false;
    }
    /**
     * Property type
     */
    get type() {
        return this._type.type;
    }
    /**
     * Returns array of decorators
     */
    getDecorators() {
        return this._decorators.slice();
    }
}
exports.PropertyInfo = PropertyInfo;
/**
 * @internal
 */
class PropertyInfoActivator extends PropertyInfo {
}
exports.PropertyInfoActivator = PropertyInfoActivator;
