// typeorm-typedi-extensions/esm2015/decorators/inject-repository.decorator.js
import { ConnectionManager, Repository, TreeRepository, MongoRepository } from 'typeorm';
import { Container } from 'typedi';
import { EntityTypeMissingError } from '../errors/entity-type-missing.error';
import { PropertyTypeMissingError } from '../errors/property-type-missing.error';
import { ParamTypeMissingError } from '../errors/param-type-missing.error';
import { ConnectionNotFoundError } from '../errors/manager-not-found.error';
/**
 * Helper to avoid V8 compilation of anonymous function on each call of decorator.
 */
function getRepositoryHelper(connectionName, repositoryType, entityType, containerInstance) {
    const connectionManager = containerInstance.get(ConnectionManager);
    if (!connectionManager.has(connectionName)) {
        throw new ConnectionNotFoundError(connectionName);
    }
    const connection = connectionManager.get(connectionName);
    switch (repositoryType) {
        case Repository:
            return connection.getRepository(entityType);
        case MongoRepository:
            return connection.getMongoRepository(entityType);
        case TreeRepository:
            return connection.getTreeRepository(entityType);
        default:
            /** If the requested type is not well-known, then it must be a custom repository. */
            return connection.getCustomRepository(repositoryType);
    }
}
export function InjectRepository(entityTypeOrConnectionName, paramConnectionName = 'default') {
    return (object, propertyName, index) => {
        let entityType;
        let connectionName;
        let repositoryType;
        // handle first parameter overload
        connectionName = paramConnectionName;
        if (typeof entityTypeOrConnectionName === 'string') {
            connectionName = entityTypeOrConnectionName;
        }
        else if (typeof entityTypeOrConnectionName === 'function') {
            entityType = entityTypeOrConnectionName;
        }
        if ((Reflect === null || Reflect === void 0 ? void 0 : Reflect.getOwnMetadata) == undefined) {
            throw new Error('Reflect.getOwnMetadata is not defined. Make sure to import the `reflect-metadata` package!');
        }
        if (index !== undefined) {
            /** The decorator has been applied to a constructor parameter. */
            const paramTypes = Reflect.getOwnMetadata('design:paramtypes', object, propertyName);
            if (!paramTypes || !paramTypes[index]) {
                throw new ParamTypeMissingError(object, propertyName, index);
            }
            repositoryType = paramTypes[index];
        }
        else {
            /** The decorator has been applied to a class property. */
            const propertyType = Reflect.getOwnMetadata('design:type', object, propertyName);
            if (!propertyType) {
                throw new PropertyTypeMissingError(object, propertyName);
            }
            repositoryType = propertyType;
        }
        switch (repositoryType) {
            case Repository:
            case MongoRepository:
            case TreeRepository:
                if (!entityType) {
                    throw new EntityTypeMissingError(object, propertyName, index);
                }
        }
        Container.registerHandler({
            object: object,
            index: index,
            propertyName: propertyName,
            value: containerInstance => getRepositoryHelper(connectionName, repositoryType, entityType, containerInstance),
        });
    };
}
