// tst-reflect/dist/type-builder/TypeBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeBuilder = void 0;
const ArrayBuilder_1 = require("./ArrayBuilder");
const IntersectionTypeBuilder_1 = require("./IntersectionTypeBuilder");
const MethodBuilder_1 = require("./MethodBuilder");
const ObjectLiteralTypeBuilder_1 = require("./ObjectLiteralTypeBuilder");
const PropertyBuilder_1 = require("./PropertyBuilder");
const UnionTypeBuilder_1 = require("./UnionTypeBuilder");
class TypeBuilder {
    constructor() {
    }
    static createUnion(types) {
        return new UnionTypeBuilder_1.UnionTypeBuilder().addTypes(...types);
    }
    static createIntersection(types) {
        return new IntersectionTypeBuilder_1.IntersectionTypeBuilder().addTypes(...types);
    }
    static createArray() {
        return new ArrayBuilder_1.ArrayTypeBuilder();
    }
    static createObject() {
        return new ObjectLiteralTypeBuilder_1.ObjectLiteralTypeBuilder();
    }
    static createProperty(description) {
        return new PropertyBuilder_1.PropertyBuilder(description);
    }
    static createMethod(description) {
        return new MethodBuilder_1.MethodBuilder(description);
    }
}
exports.TypeBuilder = TypeBuilder;
