// tst-reflect/dist/type-builder/UnionTypeBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnionTypeBuilder = void 0;
const enums_1 = require("../enums");
const Type_1 = require("../Type");
const TypeBuilderBase_1 = require("./TypeBuilderBase");
class UnionTypeBuilder extends TypeBuilderBase_1.TypeBuilderBase {
    constructor() {
        super(...arguments);
        this.types = new Set();
    }
    /**
     * Add types to union
     * @param types
     */
    addTypes(...types) {
        for (let type of types) {
            this.types.add(type);
        }
        return this;
    }
    /**
     * Build Union Type.
     * Does not return Union if there is only one type.
     * Returns Type.Undefined if there is no type.
     */
    build() {
        const types = Array.from(this.types);
        if (types.length === 0) {
            return Type_1.Type.Undefined;
        }
        if (types.length === 1) {
            return types[0];
        }
        return Type_1.Type.store.wrap({
            k: enums_1.TypeKind.Container,
            n: this.typeName,
            fn: this.fullName,
            union: true,
            types: types
        });
    }
}
exports.UnionTypeBuilder = UnionTypeBuilder;
