// tst-reflect/dist/mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mapper = void 0;
const decorator_1 = require("./descriptions/decorator");
const IndexInfo_1 = require("./descriptions/IndexInfo");
const methodInfo_1 = require("./descriptions/methodInfo");
const propertyInfo_1 = require("./descriptions/propertyInfo");
exports.Mapper = {
    /**
     * @internal
     * @param d
     */
    mapDecorators(d) {
        return Reflect.construct(decorator_1.Decorator, [d], decorator_1.DecoratorActivator);
    },
    /**
     * @internal
     * @param p
     */
    mapProperties(p) {
        return Reflect.construct(propertyInfo_1.PropertyInfo, [p], propertyInfo_1.PropertyInfoActivator);
    },
    /**
     * @internal
     * @param i
     */
    mapIndexes(i) {
        return new IndexInfo_1.IndexInfo(i);
    },
    /**
     * @internal
     * @param c
     */
    mapConstructors(c) {
        return Reflect.construct(methodInfo_1.ConstructorInfo, [c], methodInfo_1.ConstructorInfoActivator);
    },
    /**
     * @internal
     * @param m
     */
    mapMethods(m) {
        return Reflect.construct(methodInfo_1.MethodInfo, [m], methodInfo_1.MethodInfoActivator);
    }
};
